crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4240,6 +4240,46 @@ impl<'db> Class<'db> {
                     //
                     //     self.name = <value>
 
+                    // Check for a special case - unannotated assignments in `__init__` method
+                    // that assign a method param with declared type. E.g.:
+                    // ```python
+                    // class A:
+                    //     def __init__(self, name: str):
+                    //         self.name = name
+                    // ```
+                    // In this case we infer attribute type as if it had been declared with
+                    // the type of the value assigned to it, without union with Unknown.
```

Comment:
```suggestion
                    // If we see that a *declared* parameter of a method is directly assigned
                    // to an instance attribute, we treat that as if the attribute assignment had
                    // been annotated with that same parameter type. For example, we treat
                    // the following attribute assignment, as if it had been annotated with
                    // `self.name: str = name`:
                    // ```python
                    // class A:
                    //     def __init__(self, name: str):
                    //         self.name = name
                    // ```
```